src/explore-education-statistics-common/src/modules/table-tool/utils/applyTableHeadersOrder.ts

```diff
@@ -0,0 +1,439 @@
+  otherAxisLastGroupKey: LastGroupKeys;
+}
+const getKeys = (axis: Axes): Keys => {
+  return {
+    groupsKey: `${axis}Groups`,
+    lastGroupKey: `${axis}s`,
+    otherAxisGroupsKey: axis === 'row' ? 'columnGroups' : 'rowGroups',
+    otherAxisLastGroupKey: axis === 'row' ? 'columns' : 'rows',
+  };
+};
+
+const getFilterType = (filter: Filter): string => {
+  if (filter instanceof CategoryFilter) {
+    return (filter as CategoryFilter).category;
+  }
+  if (filter instanceof LocationFilter) {
+    return 'locationFilter';
+  }
+  if (filter instanceof TimePeriodFilter) {
+    return 'timePeriodFilter';
+  }
+  return 'indicator';
+};
+
+const getMatchingGroup = (
+  groups: Filter[][],
+  groupToMatch: Filter[],
+): Filter[] | undefined => {
+  const groupToMatchFilterType = getFilterType(groupToMatch[0]);
+  return groups.find(
+    group => getFilterType(group[0]) === groupToMatchFilterType,
+  );
+};
+
+/**
+ * Order items within a group based on the previous reordering.
+ * Items that weren't in the reordered list will be after the reordered ones.
+ */
+const orderItems = (
+  reorderedItems: Filter[],
+  updatedItems: Filter[],
+): Filter[] => {
+  return reorderedItems
+    .filter(item =>
+      updatedItems.find(updatedItem => updatedItem.value === item.value),
+    )
+    .concat(
+      updatedItems.filter(
+        updatedItem =>
+          !reorderedItems.find(item => item.value === updatedItem.value),
+      ),
+    );
+};
+
+const isReorderedGroup = (
+  originalGroup: Filter[],
+  updatedGroup: Filter[],
+): boolean => {
+  const intersectionOrderedByOriginal = intersectionWith(
+    originalGroup,
+    updatedGroup,
+    isEqual,
+  );
+  const intersectionOrderedByUpdated = intersectionWith(
+    updatedGroup,
+    originalGroup,
+    isEqual,
+  );
+  return !isEqual(intersectionOrderedByOriginal, intersectionOrderedByUpdated);
+};
+
+/**
+ * Have the groups in an axis been re-ordered or moved between axes
+ * */
+const areGroupsReordered = (
+  previousOrder: string[],
+  previousOtherAxisOrder: string[],
+  updatedOrder: string[],
+  updatedOtherAxisOrder: string[],
+) => {
+  if (isEqual(previousOrder, updatedOrder)) {
+    return false;
+  }
+
+  const intersectionOrderedByPrevious = intersection(
+    previousOrder,
+    updatedOrder,
+  );
+  const intersectionOrderedByUpdated = intersection(
+    updatedOrder,
+    previousOrder,
+  );
+  if (!isEqual(intersectionOrderedByPrevious, intersectionOrderedByUpdated)) {
+    return true;
+  }
+  const anyGroupsMovedOutOfAxis = previousOrder.find(group =>
+    updatedOtherAxisOrder.includes(group),
+  );
+  const anyGroupsMovedIntoAxis = updatedOrder.find(group =>
+    previousOtherAxisOrder.includes(group),
+  );
+
+  if (anyGroupsMovedOutOfAxis || anyGroupsMovedIntoAxis) {
+    return true;
+  }
+  return false;
+};
+
+const getAdditionalGroups = (
+  otherAxisReorderedGroups: Filter[][],
+  reorderedGroups: Filter[][],
+  updatedGroups: Filter[][],
+) => {
```

Comment:
Would use object syntax to make the parameters more explicit. Are they're all the same type, it'd be really easy to put one of the parameters in the wrong position and it wouldn't be noticed.

```ts
function getNewGroups({
  otherAxisReorderedGroups,
  reorderedGroups,
  defaultGroups
} : {
  otherAxisReorderedGroups: Filter[][],
  reorderedGroups: Filter[][],
  defaultGroups: Filter[][],
}): Filter[][] {
```

I'm also suggesting some additional tweaks:
- Changing the name to `getNewGroups` as it isn't too clear what 'additional' means
- Added the return type to make it clearer what this function is returning. Would generally suggest adding return types as inference makes the compiler work harder, and can also make readers have to work harder as well to figure out the return type (which is bad).